### PR TITLE
Fix Adam FP16 overflow on gpu kernels

### LINKED ISSUE #7690: Flaky test: `TestAdam`
https://jenkins.preferred.jp/job/chainer/job/chainer_pr/1490/TEST=chainer-py3,label=mn1-p100/

```
20:51:03 _ TestAdam_param_3_{adabound=False, amsgrad=False, dtype=float16, loss_scaling='static', use_placeholder=True}.test_linear_model_multi_gpu _
20:51:03 chainer/testing/parameterized.py:89: in new_method
20:51:03     utils._raise_from(e.__class__, s.getvalue(), e)
20:51:03 chainer/utils/__init__.py:106: in _raise_from
20:51:03     six.raise_from(new_exc.with_traceback(orig_exc.__traceback__), None)
20:51:03 chainer/testing/parameterized.py:78: in new_method
20:51:03     return base_method(self, *args, **kwargs)
20:51:03 chainer/testing/condition.py:80: in wrapper
20:51:03     fail()
20:51:03 chainer/testing/condition.py:51: in fail
20:51:03     instance.fail(msg)
20:51:03 E   AssertionError: Parameterized test failed.
20:51:03 E   
20:51:03 E   Base test method: TestAdam.test_linear_model_multi_gpu
20:51:03 E   Test parameters:
20:51:03 E     adabound: False
20:51:03 E     amsgrad: False
20:51:03 E     dtype: <class 'numpy.float16'>
20:51:03 E     loss_scaling: static
20:51:03 E     use_placeholder: True
20:51:03 E   
20:51:03 E   
20:51:03 E   (caused by)
20:51:03 E   AssertionError: 
20:51:03 E   Fail: 10, Success: 0
20:51:03 E   
20:51:03 E   The first error message:
20:51:03 E   Traceback (most recent call last):
20:51:03 E     File "/work/chainer/chainer/testing/condition.py", line 60, in <lambda>
20:51:03 E       lambda: f(ins, *args[1:], **kwargs),
20:51:03 E     File "/work/chainer/tests/chainer_tests/optimizers_tests/test_optimizers_by_linear_model.py", line 168, in test_linear_model_multi_gpu
20:51:03 E       self.assertGreater(cuda.to_cpu(accuracy.data), 0.9)
20:51:03 E   AssertionError: array(0.5, dtype=float16) not greater than 0.9
```

### PR DESCRIPTION
**Pull request by emcastillo**
Fix #7690

Found that it was due to lack of precision as some values were not being converted to FP32 during intermediate calculations.

### CODE REVIEW

**Comment by niboshi** [comment]
Jenkins, test this please

**Comment by niboshi** [comment]
Please check the test failures.

**Comment by emcastillo** [comment]
There was a lack of conversions back from `m`, `v` and `vhat`

File: chainer/optimizers/adam.py [hunk 1]
```diff
@@ -240,12 +240,18 @@ def update_core_gpu(self, param):
...
-                       vhat = max(vhat, v);
+                       T m_ = static_cast<T>(m);
+                       T v_ = static_cast<T>(v);
+                       T vhat_ = static_cast<T>(vhat);
+                       m_ += one_minus_beta1 * (grad_ - m_);
+                       v_ += one_minus_beta2 * (grad_ * grad_ - v_);
+                       vhat_ = max(vhat_, v_);
+                       vhat = static_cast<T>(vhat_);
+                       m = static_cast<T>(m_);
+                       v = static_cast<T>(v_);
```
Comment (niboshi) [inline chainer/optimizers/adam.py]: `vhat`, `m` and `v` have the type `P`. What's the point of those casts?

Comment (emcastillo) [reply chainer/optimizers/adam.py]: Sorry, typo. Should be casted back to P.
Im retrying the checks to see if it fails or not.
If everything goes well I will push the update later.

**Comment by niboshi** [comment]
Jenkins, test this please

File: chainer/optimizers/adam.py [hunk 2]
```diff
@@ -291,16 +307,21 @@ def update_core_gpu(self, param):
                 hp.eta, hp.weight_decay_rate, self._dummy,
                 param.data, self.state['m'], self.state['v'],
                 self.state['vhat'])
+            print(self.state['m'], self.state['v'], self.state['vhat'])
```
Comment (niboshi) [inline chainer/optimizers/adam.py]: Please remove this line.

Comment (emcastillo) [reply chainer/optimizers/adam.py]: ugh!

File: tests/chainer_tests/optimizers_tests/test_optimizers.py [hunk 3]
```diff
@@ -212,27 +212,27 @@ def test_amsgrad(self, backend_config):
         opt.update()
         testing.assert_allclose(
             x.update_rule.state['v'],
-            [0.3, 0.3, 30, 30],
+            np.array([0.3, 0.3, 30, 30]),
```
[the remark this answers is not in the document]
Comment (emcastillo) [reply tests/chainer_tests/optimizers_tests/test_optimizers.py]: If the assert_allclose fails, when trying to print the exception,
the print data fails too due to that value being a list and not an array. 
Then the original reason of the failure got lost in the output stream.

[the remark this answers is not in the document]
Comment (emcastillo) [reply tests/chainer_tests/optimizers_tests/test_optimizers.py]: The specific line of code that causes this is `chainer/testing/array.py:33`
```
   except AssertionError as e:
        f = six.StringIO()
        f.write(str(e) + '\n\n')
        f.write(
            'assert_allclose failed: \n' +
            '  shape: {} {}\n'.format(x.shape, y.shape) +
            '  dtype: {} {}\n'.format(x.dtype, y.dtype))
```
The `list` object has no `shape` attribute and all the other exception information is lost.

[the remark this answers is not in the document]
Comment (niboshi) [reply tests/chainer_tests/optimizers_tests/test_optimizers.py]: I see. Thank you for explaining.
I think it's `testing.assert_allclose` that needs a fix, not these tests.
How about reverting them from this PR?

[the remark this answers is not in the document]
Comment (emcastillo) [reply tests/chainer_tests/optimizers_tests/test_optimizers.py]: Ok!

**Comment by niboshi** [comment]
Jenkins, test this please

**Comment by niboshi** [comment]
Jenkins failure is unrelated.

**Comment by niboshi** [comment]
LGTM

**Pull request merged**